### PR TITLE
Improved Nodejs streaming inference with VAD and FFmpeg

### PR DESCRIPTION
**Pull request by igorfritzsch**

### CODE REVIEW

**Comment by lissyx** [comment]
@igorfritzsch Thanks, but please rebase and avoid merges :)

**Comment by igorfritzsch** [comment]
I'm sorry, I did not think :)

**Comment by lissyx** [comment]
@igorfritzsch Looks like we still need a rebase and merge removal, could you update your PR ?

**Comment by lissyx** [comment]
@igorfritzsch There is still a merge commit, can you please get rid of it ?

**Comment by igorfritzsch** [comment]
Sorry, tried without command line. Now should be working.

File: examples/ffmpeg_vad_streaming/index.js [hunk 1]
```diff
@@ -4,14 +4,15 @@ const VAD = require("node-vad");
...
 
 // These constants control the beam search decoder
 
 // Beam width used in the CTC decoder when building candidate transcriptions
-const BEAM_WIDTH = 1024;
+const BEAM_WIDTH = 500;
 
 // The alpha hyperparameter of the CTC decoder. Language Model weight
-const LM_ALPHA = 0.75;
+const LM_ALPHA  = 0.75;
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Could we please avoid that kind of changes ?

File: examples/ffmpeg_vad_streaming/index.js [hunk 2]
```diff
@@ -27,29 +28,29 @@ const N_FEATURES = 26;
 const N_CONTEXT = 9;
 
 let VersionAction = function VersionAction(options) {
-	options = options || {};
-	options.nargs = 0;
-	argparse.Action.call(this, options);
+  options = options || {};
+  options.nargs = 0;
+  argparse.Action.call(this, options);
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Same, no need to change indentation

File: examples/ffmpeg_vad_streaming/index.js [hunk 3]
```diff
@@ -27,29 +28,29 @@ const N_FEATURES = 26;
...
 };
 
 util.inherits(VersionAction, argparse.Action);
 
 VersionAction.prototype.call = function(parser) {
-	Ds.printVersions();
-	process.exit(0);
+  Ds.printVersions();
+  process.exit(0);
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Same here

File: examples/ffmpeg_vad_streaming/index.js [hunk 4]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
-	model.enableDecoderWithLM(args['alphabet'], args['lm'], args['trie'],
-		LM_ALPHA, LM_BETA);
-	const lm_load_end = process.hrtime(lm_load_start);
-	console.error('Loaded language model in %ds.', totalTime(lm_load_end));
+  console.error('Loading language model from files %s %s', args['lm'], args['trie']);
+  const lm_load_start = process.hrtime();
+  model.enableDecoderWithLM(args['alphabet'], args['lm'], args['trie'],
+    LM_ALPHA, LM_BETA);
+  const lm_load_end = process.hrtime(lm_load_start);
+  console.error('Loaded language model in %ds.', totalTime(lm_load_end));
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Same here

File: examples/ffmpeg_vad_streaming/index.js [hunk 5]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
+// Default initial allocation = 3 seconds := 150
+const PRE_ALLOC_FRAMES = 150;
+
+// Default is 16kHz
+const AUDIO_SAMPLE_RATE = 16000;
+
+const VAD_MODE = VAD.Mode.NORMAL;
+// const VAD_MODE = VAD.Mode.LOW_BITRATE;
+// const VAD_MODE = VAD.Mode.AGGRESSIVE;
+// const VAD_MODE = VAD.Mode.VERY_AGGRESSIVE;
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Could you add a comment explaining the differences ?

File: examples/ffmpeg_vad_streaming/index.js [hunk 6]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
+
+// Debounce time
+const DEBOUNCE_TIME = 20;
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Maybe worth commenting the use of that and the unit to help understand ; I guess it's in `ms` ?

File: examples/ffmpeg_vad_streaming/index.js [hunk 7]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
-	console.error('Running inference.');
-	console.log('Transcription: ', model.finishStream(sctx));
-	const model_load_end = process.hrtime(model_load_start);
-	console.error('Inference took %ds.', totalTime(model_load_end));
+  const model_load_start = process.hrtime();
+  console.error('Running inference.');
+  console.log('Transcription: ', model.finishStream(sctx));
+  const model_load_end = process.hrtime(model_load_start);
+  console.error('Inference took %ds for %ds audio file.', totalTime(model_load_end), audioLength.toPrecision(4));
+  audioLength = 0;
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Please avoid changing indentation

File: examples/ffmpeg_vad_streaming/index.js [hunk 8]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
+  '-hide_banner',
+  '-nostats',
+  '-loglevel', 'fatal',
+  '-i', args['audio'],
+  '-vn',
+  '-acodec', 'pcm_s16le',
+  '-ac', 1,
+  '-ar', AUDIO_SAMPLE_RATE,
+  '-f', 's16le',
+  'pipe:'
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Here as well, keep indentation

File: examples/ffmpeg_vad_streaming/index.js [hunk 9]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
-	'-vn',
-	'-acodec', 'pcm_s16le',
-	'-ac', 1,
-	'-ar', 16000,
-	'-f', 's16le',
-	'pipe:'
-]);
+function intermediateDecode() {
+  finishStream();
+  sctx = model.setupStream(PRE_ALLOC_FRAMES, AUDIO_SAMPLE_RATE);
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Keep indentation

File: examples/ffmpeg_vad_streaming/index.js [hunk 10]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
-			case VAD.Event.VOICE:
-				state = voice.START;
-				model.feedAudioContent(sctx, chunk.slice(0, chunk.length / 2));
-				break;
-		}
-	});
-});
+function feedAudioContent(chunk) {
+  audioLength += (chunk.length / 2) * ( 1 / AUDIO_SAMPLE_RATE);
+  model.feedAudioContent(sctx, chunk.slice(0, chunk.length / 2));
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Keep indentation

File: examples/ffmpeg_vad_streaming/index.js [hunk 11]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
+}
 
-ffmpeg.stdout.on('close', code => {
-	finishStream();
-});
+function processVad(data) {
+  if (data.speech.start||data.speech.state) feedAudioContent(data.audioData)
+  else if (data.speech.end) { feedAudioContent(data.audioData); intermediateDecode() }
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Keep indentation

File: examples/ffmpeg_vad_streaming/index.js [hunk 12]
```diff
@@ -59,59 +60,74 @@ const model_load_end = process.hrtime(model_load_start);
...
+function feedAudioContent(chunk) {
+  audioLength += (chunk.length / 2) * ( 1 / AUDIO_SAMPLE_RATE);
+  model.feedAudioContent(sctx, chunk.slice(0, chunk.length / 2));
+}
 
-ffmpeg.stdout.on('close', code => {
-	finishStream();
-});
+function processVad(data) {
+  if (data.speech.start||data.speech.state) feedAudioContent(data.audioData)
```
Comment (lissyx) [inline examples/ffmpeg_vad_streaming/index.js]: Please retrain on single line without `{}`, it's a time bomb. Also space around ` || `

File: examples/ffmpeg_vad_streaming/index.js (hunk 5 above)
Comment (lissyx) [reply examples/ffmpeg_vad_streaming/index.js]: :+1:

**Pull request merged**